code/__HELPERS/paths/path.dm

```diff
@@ -0,0 +1,379 @@
+
+/**
+ * Processes a path (list of turfs), removes any diagonal moves
+ *
+ * path - The path to process down
+ * pass_info - Holds all the info about what this path attempt can go through
+ * simulated_only - If we are not allowed to pass space turfs
+ * avoid - A turf to be avoided
+ */
+/proc/remove_diagonals(list/path, datum/can_pass_info/pass_info, simulated_only, turf/avoid)
+	if(length(path) < 2)
+		return path
+	var/list/modified_path = list()
+
+	for(var/i in 1 to length(path) - 1)
+		var/turf/current_turf = path[i]
+		modified_path += current_turf
+		var/turf/next_turf = path[i+1]
+		var/movement_dir = get_dir(current_turf, next_turf)
+		if(!(movement_dir & (movement_dir - 1))) //cardinal movement, no need to verify
+			continue
+		var/vertical_only = movement_dir & (NORTH|SOUTH)
+		// If we can't go directly north/south, we will first go to the side,
+		if(!CAN_STEP(current_turf,get_step(current_turf, vertical_only), simulated_only, pass_info, avoid))
+			modified_path += get_step(current_turf, movement_dir & ~vertical_only)
+		else // Otherwise, we'll first go north/south, then to the side
+			modified_path += get_step(current_turf, vertical_only)
+	modified_path += path[length(path)]
+
+	return modified_path
+
+/**
+ * For seeing if we can actually move between 2 given turfs while accounting for our access and the caller's pass_flags
+ *
+ * Assumes destinantion turf is non-dense - check and shortcircuit in code invoking this proc to avoid overhead.
+ * Makes some other assumptions, such as assuming that unless declared, non dense objects will not block movement.
+ * It's fragile, but this is VERY much the most expensive part of pathing, so it'd better be fast
+ *
+ * Arguments:
+ * * destination_turf - Where are we going from where we are?
+ * * pass_info - Holds all the info about what this path attempt can go through
+*/
+/turf/proc/LinkBlockedWithAccess(turf/destination_turf, datum/can_pass_info/pass_info)
+	if(destination_turf.x != x && destination_turf.y != y) //diagonal
+		var/in_dir = get_dir(destination_turf,src) // eg. northwest (1+8) = 9 (00001001)
+		var/first_step_direction_a = in_dir & 3 // eg. north   (1+8)&3 (0000 0011) = 1 (0000 0001)
+		var/first_step_direction_b = in_dir & 12 // eg. west   (1+8)&12 (0000 1100) = 8 (0000 1000)
+
+		for(var/first_step_direction in list(first_step_direction_a,first_step_direction_b))
+			var/turf/midstep_turf = get_step(destination_turf,first_step_direction)
+			var/way_blocked = midstep_turf.density || LinkBlockedWithAccess(midstep_turf, pass_info) || midstep_turf.LinkBlockedWithAccess(destination_turf, pass_info)
+			if(!way_blocked)
+				return FALSE
+		return TRUE
+	var/actual_dir = get_dir(src, destination_turf)
+
+	/// These are generally cheaper than looping contents so they go first
+	switch(destination_turf.pathing_pass_method)
+		// This is already assumed to be true
+		//if(TURF_PATHING_PASS_DENSITY)
+		//	if(destination_turf.density)
+		//		return TRUE
```

Comment:
does this still need to be here?

Reply:
yes. It's not done to save time but it is an assertion that we make about how the proc is called, it's kept here as a reminder that the case is covered, just elsewhere

Reply:
alright